Get-DscResourceTypeInformation.ps1

```diff
@@ -0,0 +1,133 @@
+Function Get-DscResourceTypeInformation {
+  <#
+    .SYNOPSIS
+      Collate the information about a DSC resource for building a Puppet resource_api type.
+
+    .DESCRIPTION
+      This function leverages the Get-DscResource command and the AST for DSC Resources implemented
+      in PowerShell to return additional information about the DSC Resource for building a Puppet
+      resource_api compliant Type, including retrieving help info, default values, and mandatory status.
+
+    .PARAMETER DscResource
+      The DscResourceInfo object to introspect; can be passed via the pipeline, normally retrieved
+      via calling Get-DscResource.
+
+    .PARAMETER DscResourceName
+      If not passing a full object, specify the name of the DSC Resource to retrieve and introspect.
+
+    .PARAMETER ModuleName
+      If not passing a full object, specify the module name of the the DSC Resource to retrieve and introspect.
+      Can be either a string or a hash containing the keys ModuleName and ModuleVersion.
+
+    .EXAMPLE
+      Get-DscResource -Name PSRepository | Get-DscResourceTypeInformation
+    
+    .EXAMPLE
+      Get-DscResourceTypeInformation -DscResourceName PSRepository
+
+    .NOTES
+    This function currently takes EITHER:
+
+    1. A DscResource Object, as passed by Get-DSCResource
+    2. A combo of name/module to retrieve DSC Resources from
+  #>
+  [CmdletBinding()]
+  Param(
+    [Parameter(ValueFromPipeline)]
+    [Microsoft.PowerShell.DesiredStateConfiguration.DscResourceInfo[]]$DscResource,
+    $DscResourceName,
+    $ModuleName
```

Comment:
You could use parameter sets to declare that you can us either `$DSCResource` on it's own, or use `$DscResourceName` with `$ModuleName`, but not any other combination of those. You could add data types to the last two params while you're at it.

Reply:
Good call, will refactor for this